File: visitor/oc/bytedesk-oc/ui/apis/BDUIApis.m
//
//  KFDSUIApis.m
//  bdui
//
//  Created by 萝卜丝 on 2018/7/15.
//  Copyright © 2018年 Bytedesk.com. All rights reserved.
//

#import "BDUIApis.h"
#import "BDChatKFViewController.h"
#import "BDFeedbackViewController.h"
#import "BDFaqViewController.h"

#import <SafariServices/SafariServices.h>
#import <objc/runtime.h>

static BDUIApis *sharedInstance = nil;

@implementation BDUIApis

+ (BDUIApis *)sharedInstance {
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[BDUIApis alloc] init];
    });
    return sharedInstance;
}

#pragma mark - 访客端接口

+ (void)pushWorkGroupChat:(UINavigationController *)navigationController
                withWorkGroupWid:(NSString *)wId
                       withTitle:(NSString *)title {
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    chatViewController.title = title;
    chatViewController.navigationItem.backBarButtonItem.title = @"";
    chatViewController.hidesBottomBarWhenPushed = YES;
    //
    [chatViewController initWithWorkGroupWid:wId withTitle:title withPush:YES];
    [navigationController pushViewController:chatViewController animated:YES];
}

+ (void)pushWorkGroupChat:(UINavigationController *)navigationController
                withWorkGroupWid:(NSString *)wId
                       withTitle:(NSString *)title
                      withCustom:(NSDictionary *)custom {
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    chatViewController.navigationItem.backBarButtonItem.title = @"";
    chatViewController.hidesBottomBarWhenPushed = YES;
    //
    [chatViewController initWithWorkGroupWid:wId withTitle:title withPush:YES withCustom:custom];
    [navigationController pushViewController:chatViewController animated:YES];
}

+ (void)presentWorkGroupChat:(UINavigationController *)navigationController
                   withWorkGroupWid:(NSString *)wId
                          withTitle:(NSString *)title{
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    [chatViewController initWithWorkGroupWid:wId withTitle:title withPush:NO];
    //
    UINavigationController *chatNavigationController = [[UINavigationController alloc] initWithRootViewController:chatViewController];
    [navigationController presentViewController:chatNavigationController animated:YES completion:^{
        
    }];
}

+ (void)presentWorkGroupChat:(UINavigationController *)navigationController
  withModalPresentationStyle:(UIModalPresentationStyle)modalPresentationStyle
                   withWorkGroupWid:(NSString *)wId
                          withTitle:(NSString *)title{
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    [chatViewController initWithWorkGroupWid:wId withTitle:title withPush:NO];
    //
    UINavigationController *chatNavigationController = [[UINavigationController alloc] initWithRootViewController:chatViewController];
    chatNavigationController.modalPresentationStyle = modalPresentationStyle;
    [navigationController presentViewController:chatNavigationController animated:YES completion:^{
        
    }];
}


+ (void)presentWorkGroupChat:(UINavigationController *)navigationController
                   withWorkGroupWid:(NSString *)wId
                          withTitle:(NSString *)title
                         withCustom:(NSDictionary *)custom {
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    [chatViewController initWithWorkGroupWid:wId withTitle:title withPush:NO withCustom:custom];
    //
    UINavigationController *chatNavigationController = [[UINavigationController alloc] initWithRootViewController:chatViewController];
    chatNavigationController.modalPresentationStyle = UIModalPresentationFullScreen;
    [navigationController presentViewController:chatNavigationController animated:YES completion:^{
        
    }];
}


+ (void)pushAppointChat:(UINavigationController *)navigationController
                  withAgentUid:(NSString *)uId
                     withTitle:(NSString *)title {
    //
//    [BDConfig switchToKF];
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    chatViewController.navigationItem.backBarButtonItem.title = @"";
    chatViewController.hidesBottomBarWhenPushed = YES;
    //
    [chatViewController initWithAgentUid:uId withTitle:title withPush:YES];
    [navigationController pushViewController:chatViewController animated:YES];
}

+ (void)pushAppointChat:(UINavigationController *)navigationController
                  withAgentUid:(NSString *)uId
                     withTitle:(NSString *)title
                    withCustom:(NSDictionary *)custom {
    //
//    [BDConfig switchToKF];
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    chatViewController.navigationItem.backBarButtonItem.title = @"";
    chatViewController.hidesBottomBarWhenPushed = YES;
    //
    [chatViewController initWithAgentUid:uId withTitle:title withPush:YES withCustom:custom];
    [navigationController pushViewController:chatViewController animated:YES];
}

+ (void)presentAppointChat:(UINavigationController *)navigationController
                     withAgentUid:(NSString *)uId
                        withTitle:(NSString *)title {
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    [chatViewController initWithAgentUid:uId withTitle:title withPush:NO];
    //
    UINavigationController *chatNavigationController = [[UINavigationController alloc] initWithRootViewController:chatViewController];
//    chatNavigationController.modalPresentationStyle = UIModalPresentationFullScreen;
    [navigationController presentViewController:chatNavigationController animated:YES completion:^{
        
    }];
}

+ (void)presentAppointChat:(UINavigationController *)navigationController
withModalPresentationStyle:(UIModalPresentationStyle)modalPresentationStyle
                     withAgentUid:(NSString *)uId
                 withTitle:(NSString *)title {
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    [chatViewController initWithAgentUid:uId withTitle:title withPush:NO];
    //
    UINavigationController *chatNavigationController = [[UINavigationController alloc] initWithRootViewController:chatViewController];
    chatNavigationController.modalPresentationStyle = modalPresentationStyle;
    [navigationController presentViewController:chatNavigationController animated:YES completion:^{
        
    }];
}

+ (void)presentAppointChat:(UINavigationController *)navigationController
                     withAgentUid:(NSString *)uId
                        withTitle:(NSString *)title
                       withCustom:(NSDictionary *)custom {
    //
    BDChatKFViewController *chatViewController = [[BDChatKFViewController alloc] init];
    [chatViewController initWithAgentUid:uId withTitle:title withPush:NO withCustom:custom];
    //
    UINavigationController *chatNavigationController = [[UINavigationController alloc] initWithRootViewController:chatViewController];
//    chatNavigationController.modalPresentationStyle = UIModalPresentationFullScreen;
    [navigationController presentViewController:chatNavigationController animated:YES completion:^{
        
    }];
}

+ (void)pushFeedback:(UINavigationController *)navigationController withAdminUid:(NSString *)uid {
    //
    BDFeedbackViewController *feedbackViewController = [[BDFeedbackViewController alloc] init];
    feedbackViewController.hidesBottomBarWhenPushed = YES;
//   TODO: 待处理
    [feedbackViewController initWithUid:uid withPush:YES];
    //
    [navigationController pushViewController:feedbackViewController animated:YES];
}

+ (void)presentFeedback:(UINavigationController *)navigationController withAdminUid:(NSString *)uid {
    //
    BDFeedbackViewController *feedbackViewController = [[BDFeedbackViewController alloc] init];
//    [feedbackViewController initWithUid:uid withPush:NO];
    //   TODO: 待处理
    [feedbackViewController initWithUid:uid withPush:NO];
    //
    UINavigationController *fbnavigationController = [[UINavigationController alloc] initWithRootViewController:feedbackViewController];
    [navigationController presentViewController:fbnavigationController animated:YES completion:^{
        
    }];
}

+ (void)presentFeedback:(UINavigationController *)navigationController withModalPresentationStyle:(UIModalPresentationStyle)modalPresentationStyle withAdminUid:(NSString *)uid {
    //
    BDFeedbackViewController *feedbackViewController = [[BDFeedbackViewController alloc] init];
    feedbackViewController.modalPresentationStyle = modalPresentationStyle;
    [feedbackViewController initWithUid:uid withPush:NO];
    //   TODO: 待处理
//    [feedbackViewController initWithType:@"" withUid:uid withPush:NO];
    //
    UINavigationController *fbnavigationController = [[UINavigationController alloc] initWithRootViewController:feedbackViewController];
    [navigationController presentViewController:fbnavigationController animated:YES completion:^{
        
    }];
}

+ (void)pushFaqApi:(UINavigationController *)navigationController withAdminUid:(NSString *)uid {
    //
    BDFaqViewController *supportViewController = [[BDFaqViewController alloc] init];
    supportViewController.hidesBottomBarWhenPushed = YES;
//    [supportViewController initWithUid:uid withPush:YES];
    //   TODO: 待处理
    [supportViewController initWithType:@"" withUid:uid withPush:YES];
    //
    [navigationController pushViewController:supportViewController animated:YES];
}

+ (void)presentSupportURL:(UINavigationController *)navigationController withAdminUid:(NSString *)uid {
    //
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"https://vip.docs.bytedesk.com/support?uid=%@&ph=ph", uid]];
    SFSafariViewController *safariVC = [[SFSafariViewController alloc] initWithURL:url];
//    safariVC.delegate = self;
//    [navigationController pushViewController:safariVC animated:YES];
    // 建议
    [navigationController presentViewController:safariVC animated:YES completion:^{
    }];
}

#pragma mark - 截屏

// TODO: 1.可拖动
// TODO: 2.可X关闭
// TODO: 3.用户无操作，5秒后自动消失
// TODO: 4.联系客服
// TODO: 5.反馈问题
// TODO: 6.分享截图
- (void)showScreenshot:(UIWindow *)window
   withBackgroundColor:(UIColor *)backgroundColor
      withWorkGroupWid:(NSString *)workGroupWid
          withShowKeFu:(BOOL)showKefu
      withShowFeedback:(BOOL)showFeedback
         withShowShare:(BOOL)showShare
      withKefuCallback:(ScreenshotCallbackBlock)kefuCallback
  withFeedbackCallback:(ScreenshotCallbackBlock)feedbackCallback
     withShareCallback:(ScreenshotCallbackBlock)shareCallback {
    
    //人为截屏, 模拟用户截屏行为, 获取所截图片
    UIImage *imageScreenShoot = [self imageWithScreenshot];
    
    int width = 150;
    int height = 250;
    
    //添加显示
    CGRect screenshotFrame = CGRectMake(window.frame.size.width - width - 5, window.frame.size.height/2 - height/2, width, height);
    UIView *screenshotView = [[UIView alloc] initWithFrame:screenshotFrame];
    [screenshotView setBackgroundColor:backgroundColor];
    screenshotView.layer.cornerRadius = 10;
    
    // 截图
    UIImageView *screenShootImageView = [[UIImageView alloc] initWithImage:imageScreenShoot];
    [screenShootImageView setFrame:CGRectMake(10, 10, 100, 105)];
    [screenshotView addSubview:screenShootImageView];
    
    UIButton *closeButton = [[UIButton alloc] initWithFrame:CGRectMake(95, 5, 20, 20)];
    [closeButton setTitle:@"X" forState:UIControlStateNormal];
    [closeButton addTarget:self action:@selector(closeScreenShoot:) forControlEvents:UIControlEventTouchUpInside];
    [screenshotView addSubview:closeButton];
    objc_setAssociatedObject(closeButton, @"closeButton", screenshotView, OBJC_ASSOCIATION_RETAIN_NONATOMIC);

    if (showKefu) {
        UIButton *kefuButton = [[UIButton alloc] initWithFrame:CGRectMake(10, 120, 100, 20)];
        kefuButton.titleLabel.font = [UIFont systemFontOfSize:15];
        [kefuButton setTitle:@"联系客服" forState:UIControlStateNormal];
        [kefuButton setBackgroundColor:[UIColor grayColor]];
        [kefuButton addTarget:self action:@selector(kefuScreenShoot) forControlEvents:UIControlEventTouchUpInside];
        [screenshotView addSubview:kefuButton];
    }

    if (showFeedback) {
        UIButton *feedbackButton = [[UIButton alloc] initWithFrame:CGRectMake(10, 145, 100, 20)];
        feedbackButton.titleLabel.font = [UIFont systemFontOfSize:15];
        [feedbackButton setTitle:@"意见反馈" forState:UIControlStateNormal];
        [feedbackButton setBackgroundColor:[UIColor grayColor]];
        [feedbackButton addTarget:self action:@selector(feedbackScreenShoot) forControlEvents:UIControlEventTouchUpInside];
        [screenshotView addSubview:feedbackButton];
    }
    
    if (showShare) {
        UIButton *shareButton = [[UIButton alloc] initWithFrame:CGRectMake(10, 170, 100, 20)];
        shareButton.titleLabel.font = [UIFont systemFontOfSize:15];
        [shareButton setTitle:@"分享截图" forState:UIControlStateNormal];
        [shareButton setBackgroundColor:[UIColor grayColor]];
        [shareButton addTarget:self action:@selector(shareScreenShoot) forControlEvents:UIControlEventTouchUpInside];
        [screenshotView addSubview:shareButton];
    }
    //
    [window addSubview:screenshotView];
}

- (void)closeScreenShoot:(UIButton *)sender {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    UIView *feedbackView = objc_getAssociatedObject(sender, @"closeButton");
    [feedbackView removeFromSuperview];
}

- (void)kefuScreenShoot {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)feedbackScreenShoot {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)shareScreenShoot {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

// 截取当前屏幕
- (NSData *)dataWithScreenshotInPNGFormat {
    CGSize imageSize = CGSizeZero;
    UIInterfaceOrientation orientation = [UIApplication sharedApplication].windows.firstObject.windowScene.interfaceOrientation;
//    UIInterfaceOrientation orientation = [UIApplication sharedApplication].statusBarOrientation;
    if (UIInterfaceOrientationIsPortrait(orientation))
        imageSize = [UIScreen mainScreen].bounds.size; // 竖屏
    else
        imageSize = CGSizeMake([UIScreen mainScreen].bounds.size.height, [UIScreen mainScreen].bounds.size.width); // 横屏
    
    UIGraphicsBeginImageContextWithOptions(imageSize, NO, 0);
    CGContextRef context = UIGraphicsGetCurrentContext();
    for (UIWindow *window in [[UIApplication sharedApplication] windows]) {
        //
        CGContextSaveGState(context);
        CGContextTranslateCTM(context, window.center.x, window.center.y);
        CGContextConcatCTM(context, window.transform);
        CGContextTranslateCTM(context, -window.bounds.size.width * window.layer.anchorPoint.x, -window.bounds.size.height * window.layer.anchorPoint.y);
        if (orientation == UIInterfaceOrientationLandscapeLeft) {
            CGContextRotateCTM(context, M_PI_2);
            CGContextTranslateCTM(context, 0, -imageSize.width);
        }
        else if (orientation == UIInterfaceOrientationLandscapeRight) {
            CGContextRotateCTM(context, -M_PI_2);
            CGContextTranslateCTM(context, -imageSize.height, 0);
        } else if (orientation == UIInterfaceOrientationPortraitUpsideDown) {
            CGContextRotateCTM(context, M_PI);
            CGContextTranslateCTM(context, -imageSize.width, -imageSize.height);
        }
        if ([window respondsToSelector:@selector(drawViewHierarchyInRect:afterScreenUpdates:)]) {
            [window drawViewHierarchyInRect:window.bounds afterScreenUpdates:YES];
        }
        else {
            [window.layer renderInContext:context];
        }
        CGContextRestoreGState(context);
    }
    //
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return UIImagePNGRepresentation(image);
}

/**
 *  返回截取到的图片
 *
 *  @return UIImage *
 */
- (UIImage *)imageWithScreenshot {
    NSData *imageData = [self dataWithScreenshotInPNGFormat];
    return [UIImage imageWithData:imageData];
}


#pragma mark - 公共接口

+ (void) showTipWithVC:(UIViewController *)viewController  withMessage:(NSString *)message {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:nil
                                                                  message:@""
                                                           preferredStyle:UIAlertControllerStyleAlert];
    UIView *firstSubview = alertController.view.subviews.firstObject;
    UIView *alertContentView = firstSubview.subviews.firstObject;
    for (UIView *subSubView in alertContentView.subviews) {
        subSubView.backgroundColor = [UIColor colorWithRed:141/255.0f green:0/255.0f blue:254/255.0f alpha:1.0f];
    }
    //
    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] initWithString:message];
    [attributedString addAttribute:NSForegroundColorAttributeName value:[UIColor whiteColor] range:NSMakeRange(0,attributedString.length)];
    [alertController setValue:attributedString forKey:@"attributedTitle"];
    //
    [viewController presentViewController:alertController animated:YES completion:nil];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [alertController dismissViewControllerAnimated:YES completion:^{
        }];
    });
}

+ (void) showErrorWithVC:(UIViewController *)viewController  withMessage:(NSString *)message {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:nil
                                                                  message:@""
                                                           preferredStyle:UIAlertControllerStyleAlert];
    UIView *firstSubview = alertController.view.subviews.firstObject;
    UIView *alertContentView = firstSubview.subviews.firstObject;
    for (UIView *subSubView in alertContentView.subviews) {
        subSubView.backgroundColor = [UIColor colorWithRed:141/255.0f green:0/255.0f blue:254/255.0f alpha:1.0f];
    }
    //
    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] initWithString:message];
    [attributedString addAttribute:NSForegroundColorAttributeName value:[UIColor whiteColor] range:NSMakeRange(0,attributedString.length)];
    [alertController setValue:attributedString forKey:@"attributedTitle"];
    //
    [viewController presentViewController:alertController animated:YES completion:nil];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [alertController dismissViewControllerAnimated:YES completion:^{
        }];
    });
}

+ (void)screenshot {
    
    UIWindow *keyWindow = [UIApplication sharedApplication].windows.firstObject; //[[UIApplication sharedApplication] keyWindow];
    
    CGRect rect = [keyWindow bounds];
    UIGraphicsBeginImageContextWithOptions(rect.size, YES, 0);
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    [keyWindow.layer renderInContext:context];
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    [[UIPasteboard generalPasteboard] setImage:image];
}


@end

File: visitor/swift/Sources/bytedesk-swift/Apis/BDUIApis.swift
//
//  File.swift
//  
//
//  Created by 宁金鹏 on 2023/8/19.
//

import UIKit
import ObjectiveC.runtime

// Swift 中的类如果要供Objective-C 调用，必须也继承自NSObject
public class BDUIApis: NSObject {
    //
    public class func sharedInstance() -> BDUIApis {
        struct Static {
            static let instance = BDUIApis()
        }
        return Static.instance
    }
    
    static public func pushWorkGroupChat(navController: UINavigationController, wid: String, title: String) {
        let chatVC = BDChatKFViewController()
        chatVC.initWithWorkGroupWid(wid: wid, title, true)
        navController.pushViewController(chatVC, animated: true)
    }
    
    static public func pushWorkGroupChat(navController: UINavigationController, wid: String, title: String, custom: [String: Any]) {
        let chatVC = BDChatKFViewController()
        chatVC.initWithWorkGroupWid(wid: wid, title, true, custom: custom)
        navController.pushViewController(chatVC, animated: true)
    }
    
    static public func presentWorkGroupChat(navController: UINavigationController, wid: String, title: String) {
        let chatVC = BDChatKFViewController()
        chatVC.initWithWorkGroupWid(wid: wid, title, false)
        let chatNavigationController = UINavigationController.init(rootViewController: chatVC)
        navController.present(chatNavigationController, animated: true)
    }
    
    static public func presentWorkGroupChat(navController: UINavigationController, wid: String, title: String, custom: [String: Any]) {
        let chatVC = BDChatKFViewController()
        chatVC.initWithWorkGroupWid(wid: wid, title, false, custom: custom)
        let chatNavigationController = UINavigationController.init(rootViewController: chatVC)
        navController.present(chatNavigationController, animated: true)
    }
    
    static public func pushAppointChat(navController: UINavigationController, uid: String, title: String) {
        let chatVC = BDChatKFViewController()
        chatVC.initWithAgentUid(uid: uid, title, true)
        navController.pushViewController(chatVC, animated: true)
    }
    
    static public func pushAppointChat(navController: UINavigationController, uid: String, title: String, custom: [String: Any]) {
        let chatVC = BDChatKFViewController()
        chatVC.initWithAgentUid(uid: uid, title, true, custom: custom)
        navController.pushViewController(chatVC, animated: true)
    }
    
    static public func presentAppointChat(navController: UINavigationController, uid: String, title: String) {
        let chatVC = BDChatKFViewController()
        chatVC.initWithAgentUid(uid: uid, title, false)
        let chatNavigationController = UINavigationController.init(rootViewController: chatVC)
        navController.present(chatNavigationController, animated: true)
    }
    
    static public func presentAppointChat(navController: UINavigationController, uid: String, title: String, custom: [String: Any]) {
        let chatVC = BDChatKFViewController()
        chatVC.initWithAgentUid(uid: uid, title, false, custom: custom)
        let chatNavigationController = UINavigationController.init(rootViewController: chatVC)
        navController.present(chatNavigationController, animated: true)
    }
    
    static public func pushFeedback(navController: UINavigationController, adminUid: String, title: String) {
        let feedbackVC = BDFeedbackViewController()
        feedbackVC.initWithUid(adminUid, withPush: true)
        navController.pushViewController(feedbackVC, animated: true)
    }
    
    static public func presentFeedback(navController: UINavigationController, adminUid: String, title: String) {
        let feedbackVC = BDFeedbackViewController()
        feedbackVC.initWithUid(adminUid, withPush: false)
        let feedbackNavigationController = UINavigationController.init(rootViewController: feedbackVC)
        navController.present(feedbackNavigationController, animated: true)
    }
    
    static public func pushWebView(navController: UINavigationController, chatUrl: String, title: String) {
        let webVC = BDWebViewViewController()
        webVC.initWithChatUrl(chatUrl, isPush: true)
        navController.pushViewController(webVC, animated: true)
    }
    
    static public func presentWebView(navController: UINavigationController, chatUrl: String, title: String) {
        let webVC = BDWebViewViewController()
        webVC.initWithChatUrl(chatUrl, isPush: false)
        let webNavigationController = UINavigationController.init(rootViewController: webVC)
        navController.present(webNavigationController, animated: true)
    }
    
    // MARK: 截屏反馈
    public typealias ScreenshotCallbackBlock = () -> Void
    
    public func showScreenshot(window: UIWindow,
                        backgroundColor: UIColor,
                        workGroupWid: String,
                        showKefu: Bool,
                        showFeedback: Bool,
                        showShare: Bool,
                        kefuCallback: ScreenshotCallbackBlock?,
                        feedbackCallback: ScreenshotCallbackBlock?,
                        shareCallback: ScreenshotCallbackBlock?) {
        
        // 人为截屏, 模拟用户截屏行为, 获取所截图片
        let imageScreenShoot = imageWithScreenshot()
        
        let width = 150
        let height = 250
        
        // 添加显示
        let screenshotFrame = CGRect(x: window.frame.size.width - CGFloat(width) - 5,
                                     y: window.frame.size.height/2 - CGFloat(height)/2,
                                     width: CGFloat(width),
                                     height: CGFloat(height))
        let screenshotView = UIView(frame: screenshotFrame)
        screenshotView.backgroundColor = backgroundColor
        screenshotView.layer.cornerRadius = 10
        
        // 截图
        let screenShootImageView = UIImageView(image: imageScreenShoot)
        screenShootImageView.frame = CGRect(x: 10, y: 10, width: 100, height: 105)
        screenshotView.addSubview(screenShootImageView)
        
        let closeButton = UIButton(frame: CGRect(x: width - 20, y: 5, width: 20, height: 20))
        closeButton.setImage(UIImage(named: "xmark", in: .module, with: nil), for: .normal)
        closeButton.setTitleColor(UIColor.white, for: .normal)
        closeButton.addTarget(self, action: #selector(closeScreenShoot(_:)), for: .touchUpInside)
        screenshotView.addSubview(closeButton)
        objc_setAssociatedObject(closeButton, "closeButton", screenshotView, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        
        if showKefu {
            let kefuButton = UIButton(frame: CGRect(x: 10, y: 120, width: 100, height: 20))
            kefuButton.titleLabel?.font = UIFont.systemFont(ofSize: 15)
            kefuButton.setTitle("联系客服", for: .normal)
            kefuButton.backgroundColor = UIColor.gray
            kefuButton.addTarget(self, action: #selector(kefuScreenShoot), for: .touchUpInside)
            screenshotView.addSubview(kefuButton)
        }
        
        if showFeedback {
            let feedbackButton = UIButton(frame: CGRect(x: 10, y: 145, width: 100, height: 20))
            feedbackButton.titleLabel?.font = UIFont.systemFont(ofSize: 15)
            feedbackButton.setTitle("意见反馈", for: .normal)
            feedbackButton.backgroundColor = UIColor.gray
            feedbackButton.addTarget(self, action: #selector(feedbackScreenShoot), for: .touchUpInside)
            screenshotView.addSubview(feedbackButton)
        }
        
        if showShare {
            let shareButton = UIButton(frame: CGRect(x: 10, y: 170, width: 100, height: 20))
            shareButton.titleLabel?.font = UIFont.systemFont(ofSize: 15)
            shareButton.setTitle("分享截图", for: .normal)
            shareButton.backgroundColor = UIColor.gray
            shareButton.addTarget(self, action: #selector(shareScreenShoot), for: .touchUpInside)
            screenshotView.addSubview(shareButton)
        }
        
        window.addSubview(screenshotView)
    }
    
    @objc func closeScreenShoot(_ sender: UIButton) {
        print(#function)
        if let feedbackView = objc_getAssociatedObject(sender, "closeButton") as? UIView {
            feedbackView.removeFromSuperview()
        }
    }
    
    @objc func kefuScreenShoot() {
        print(#function)
    }
    
    @objc func feedbackScreenShoot() {
        print(#function)
    }
    
    @objc func shareScreenShoot() {
        print(#function)
    }
    
    // 截取当前屏幕
    func dataWithScreenshotInPNGFormat() -> Data? {
        var imageSize = CGSize.zero
        let orientation = UIApplication.shared.windows.first?.windowScene?.interfaceOrientation ?? .unknown
        
        if orientation.isPortrait {
            imageSize = UIScreen.main.bounds.size // 竖屏
        } else {
            imageSize = CGSize(width: UIScreen.main.bounds.size.height,
                               height: UIScreen.main.bounds.size.width) // 横屏
        }
        
        UIGraphicsBeginImageContextWithOptions(imageSize, false, 0)
        guard let context = UIGraphicsGetCurrentContext() else {
            return nil
        }
        
        for window in UIApplication.shared.windows {
            context.saveGState()
            context.translateBy(x: window.center.x, y: window.center.y)
            context.concatenate(window.transform)
            context.translateBy(x: -window.bounds.size.width * window.layer.anchorPoint.x,
                                y: -window.bounds.size.height * window.layer.anchorPoint.y)
                
            if orientation == .landscapeLeft {
                context.rotate(by: CGFloat.pi/2)
                context.translateBy(x: 0, y: -imageSize.width)
            } else if orientation == .landscapeRight {
                context.rotate(by: -CGFloat.pi/2)
                context.translateBy(x: -imageSize.height, y: 0)
            } else if orientation == .portraitUpsideDown {
                context.rotate(by: CGFloat.pi)
                context.translateBy(x: -imageSize.width, y: -imageSize.height)
            }
            
            if window.responds(to: #selector(UIView.drawHierarchy(in:afterScreenUpdates:))) {
                window.drawHierarchy(in: window.bounds, afterScreenUpdates: true)
            } else {
                window.layer.render(in: context)
            }
            
            context.restoreGState()
        }
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return image?.pngData()
    }
    
    // 返回截取到的图片
    func imageWithScreenshot() -> UIImage? {
        if let imageData = dataWithScreenshotInPNGFormat() {
            return UIImage(data: imageData)
        }
        return nil
    }
    
    
    // MARK: 工具函数
    
    static func showTip(with viewController: UIViewController, message: String) {
        let alertController = UIAlertController(title: nil, message: "", preferredStyle: .alert)
        if let firstSubview = alertController.view.subviews.first,
            let alertContentView = firstSubview.subviews.first {
            for subSubView in alertContentView.subviews {
                subSubView.backgroundColor = UIColor(red: 141/255.0, green: 0/255.0, blue: 254/255.0, alpha: 1.0)
            }
        }
        let attributedString = NSMutableAttributedString(string: message)
        attributedString.addAttributes([NSAttributedString.Key.foregroundColor: UIColor.white], range: NSRange(location: 0, length: attributedString.length))
        alertController.setValue(attributedString, forKey: "attributedTitle")
        viewController.present(alertController, animated: true, completion: nil)
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
            alertController.dismiss(animated: true, completion: nil)
        }
    }

    static func showError(with viewController: UIViewController, message: String) {
        let alertController = UIAlertController(title: nil, message: "", preferredStyle: .alert)
        if let firstSubview = alertController.view.subviews.first,
            let alertContentView = firstSubview.subviews.first {
            for subSubView in alertContentView.subviews {
                subSubView.backgroundColor = UIColor(red: 141/255.0, green: 0/255.0, blue: 254/255.0, alpha: 1.0)
            }
        }
        let attributedString = NSMutableAttributedString(string: message)
        attributedString.addAttributes([NSAttributedString.Key.foregroundColor: UIColor.white], range: NSRange(location: 0, length: attributedString.length))
        alertController.setValue(attributedString, forKey: "attributedTitle")
        viewController.present(alertController, animated: true, completion: nil)
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            alertController.dismiss(animated: true, completion: nil)
        }
    }
    
}
